src/taker/api2.rs

```diff
@@ -0,0 +1,1376 @@
+    pub last_maker_sender_nonce: Option<crate::protocol::messages2::SerializablePublicNonce>,
+}
+
+pub(crate) const TCP_TIMEOUT_SECONDS: u64 = 300;
+pub(crate) const REFUND_LOCKTIME: u16 = 20;
+pub(crate) const REFUND_LOCKTIME_STEP: u16 = 20;
+
+/// Helper function to establish a connection to a maker
+fn connect_to_maker(
+    maker_addr: &str,
+    config: &TakerConfig,
+    timeout_secs: u64,
+) -> Result<TcpStream, TakerError> {
+    let socket = match config.connection_type {
+        ConnectionType::CLEARNET => {
+            TcpStream::connect(maker_addr)
+                .map_err(|e| TakerError::General(format!("Failed to connect to maker {}: {:?}", maker_addr, e)))?
+        }
+        ConnectionType::TOR => {
+            Socks5Stream::connect(
+                format!("127.0.0.1:{}", config.socks_port).as_str(),
+                maker_addr,
+            )
+            .map_err(|e| TakerError::General(format!("Failed to connect to maker {} via Tor: {:?}", maker_addr, e)))?
+            .into_inner()
+        }
+    };
+
+    let timeout = Duration::from_secs(timeout_secs);
+    socket.set_read_timeout(Some(timeout))
+        .map_err(|e| TakerError::General(format!("Failed to set read timeout: {:?}", e)))?;
+    socket.set_write_timeout(Some(timeout))
+        .map_err(|e| TakerError::General(format!("Failed to set write timeout: {:?}", e)))?;
+
+    Ok(socket)
+}
+
+/// Send a message with 0x01 prefix for taproot makers
+fn send_message_with_prefix(
+    socket_writer: &mut TcpStream,
+    message: &impl serde::Serialize,
+) -> Result<(), TakerError> {
+    use std::io::{BufWriter, Write};
+
+    let mut writer = BufWriter::new(socket_writer);
+    let mut msg_bytes = Vec::new();
+    msg_bytes.push(0x01);
+    msg_bytes.extend(serde_cbor::to_vec(message)?);
+    let msg_len = (msg_bytes.len() as u32).to_be_bytes();
+    let mut to_send = Vec::with_capacity(msg_bytes.len() + msg_len.len());
+    to_send.extend(msg_len);
+    to_send.extend(msg_bytes);
+    writer.write_all(&to_send)?;
+    writer.flush()?;
+    Ok(())
+}
+
+/// Fetch offers from taproot makers using taproot protocol messages
+fn fetch_taproot_offers(
+    maker_addresses: Vec<MakerAddress>,
+    config: &TakerConfig,
+) -> Result<Vec<OfferAndAddress>, TakerError> {
+    let mut offers = Vec::new();
+
+    for address in maker_addresses {
+        match download_taproot_offer(&address, config) {
+            Some(offer) => offers.push(offer),
+            None => {
+                log::warn!("Failed to download offer from taproot maker: {}", address);
+            }
+        }
+    }
+
+    Ok(offers)
+}
+
+/// Download a single offer from a taproot maker
+fn download_taproot_offer(address: &MakerAddress, config: &TakerConfig) -> Option<OfferAndAddress> {
+    let maker_addr = address.to_string();
+    log::info!("Downloading offer from taproot maker: {}", maker_addr);
```

Comment:
The current downloading routine, has retry clause added in it. We should use the same. Even better if we can use the existing download routine. Only the message type changes but the remaining infra should remain same.